### PR TITLE
gh-112281: Allow `Union` with unhashable `Annotated` metadata

### PR DESCRIPTION
**Pull request by sobolevn**
This should not affect performance that much:
1. I tried to keep the fast path as it is
2. Slow path is only used when `TypeError` happens on unhashable items (it should not affect code with only hashable parts)

* Issue: gh-112281

### CODE REVIEW

File: Lib/typing.py [hunk 1]
```diff
@@ -1533,7 +1556,10 @@ def copy_with(self, params):
     def __eq__(self, other):
         if not isinstance(other, (_UnionGenericAlias, types.UnionType)):
             return NotImplemented
-        return set(self.__args__) == set(other.__args__)
+        try:  # fast path
+            return set(self.__args__) == set(other.__args__)
+        except TypeError:  # not hashable, slow path
+            return _deduplicate(self.__args__) == _deduplicate(other.__args__)
```
Comment (faph) [inline Lib/typing.py]: Would this second code path not compare **in order**, while the set-approach does not?

Comment (sobolevn) [reply Lib/typing.py]: Indeed, I also don't like the assumption that only `Annotated` can be unhashable. Let's make it better.

File: Lib/test/test_typing.py [hunk 2]
```diff
@@ -8267,6 +8267,65 @@ def test_flatten(self):
...
+        self.assertEqual(get_args(Union[Annotated[int, 1], Annotated[str, 1], int]),
+                         (Annotated[int, 1], Annotated[str, 1], int))
+
+        # Duplicates:
+        self.assertEqual(Annotated[int, 1] | Annotated[int, 1] | int,
+                         Annotated[int, 1] | int)
+        self.assertEqual(Union[Annotated[int, 1], Annotated[int, 1], int],
+                         Union[Annotated[int, 1], int])
+
+        # Unhashable metdata:
```
Comment (AlexWaygood) [inline Lib/test/test_typing.py]:
```suggestion
        # Unhashable metadata:
```

File: Lib/typing.py [hunk 3]
```diff
@@ -308,19 +308,67 @@ def _unpack_args(args):
...
+            raise
+
+        # Happens for cases like `Annotated[dict, {'x': IntValidator()}]`
+        hasable, unhashable = _deduplicate_unhashable(params)
+        return hasable + unhashable
+
+    if len(all_params) == len(params):  # fast case
+        return params
+    new_params = []
+    for t in params:
```
Comment (JelleZijlstra) [inline Lib/typing.py]: Can't this be done faster with `list(dict.fromkeys(params))`?

File: Lib/typing.py [hunk 4]
```diff
@@ -308,19 +308,67 @@ def _unpack_args(args):
...
+    for t in regular_params:
+        if t in unique_params:
+            new_hashable.append(t)
+            unique_params.remove(t)
+    assert not unique_params, unique_params
+
+    new_unhashable = []
+    for t in unhashable_params:
+        # This is slow, but this is the only safe way to compare
+        # two `Annotated[]` instances, since `metadata` might be unhashable.
```
Comment (JelleZijlstra) [inline Lib/typing.py]: This comment seems inapposite, as this path isn't only for Annotated, and your tests don't even use Annotated.

File: Lib/test/test_typing.py [hunk 5]
```diff
@@ -8267,6 +8279,78 @@ def test_flatten(self):
...
+                         (Annotated[int, 1], Annotated[str, 1], int))
+
+        # Duplicates:
+        self.assertEqual(Annotated[int, 1] | Annotated[int, 1] | int,
+                         Annotated[int, 1] | int)
+        self.assertEqual(Union[Annotated[int, 1], Annotated[int, 1], int],
+                         Union[Annotated[int, 1], int])
+
+        # Unhashable metadata:
+        self.assertEqual(get_args(str | Annotated[int, {}] | Annotated[int, set()] | int),
```
Comment (JelleZijlstra) [inline Lib/test/test_typing.py:8568]: Wouldn't it be better to preserve the ordering?

File: Lib/test/test_typing.py [hunk 6]
```diff
@@ -8267,6 +8279,78 @@ def test_flatten(self):
...
+                         Annotated[int, {}] | int)
+        self.assertEqual(Annotated[int, {}] | Annotated[int, {}] | int,
+                         int | Annotated[int, {}])
+        self.assertEqual(Union[Annotated[int, {}], Annotated[int, {}], int],
+                         Union[Annotated[int, {}], int])
+        self.assertEqual(Union[Annotated[int, {}], Annotated[int, {}], int],
+                         Union[int, Annotated[int, {}]])
+
+    def test_order_in_union(self):
+        import operator, functools
```
Comment (JelleZijlstra) [inline Lib/test/test_typing.py]: Why not just import these at the top of the file? functools is already imported there.

File: Misc/NEWS.d/next/Library/2023-11-20-16-15-44.gh-issue-112281.gH4EVk.rst [hunk 7]
```diff
@@ -0,0 +1,2 @@
+Allow creating :ref:`union of types<types-union>` for
+:class:`typing.Annotated` with unhashable / mutable metadata.
```
Comment (JelleZijlstra) [inline Misc/NEWS.d/next/Library/2023-11-20-16-15-44.gh-issue-112281.gH4EVk.rst]:
```suggestion
:class:`typing.Annotated` with unhashable metadata.
```

File: Lib/typing.py [hunk 8]
```diff
@@ -308,19 +308,44 @@ def _unpack_args(args):
             newargs.append(arg)
     return newargs
 
-def _deduplicate(params):
+def _deduplicate(params, *, unhashable_fallback=True):
```
Comment (AlexWaygood) [inline Lib/typing.py]: I'm worried that there's no way of avoiding abysmal performance of `_deduplicate` for unhashable elements. Could we make the default `False` for this parameter?

File: Lib/typing.py (hunk 3 above)
Comment (AlexWaygood) [reply Lib/typing.py]: I don't think we even need to convert it to a `list` before returning it: if you look at the callsites, this function just needs to return a deduped iterable that retains the order of the input. No tests fail if I apply this diff to @sobolevn's PR branch:

```diff
--- a/Lib/typing.py
+++ b/Lib/typing.py
@@ -311,7 +311,7 @@ def _unpack_args(args):
 def _deduplicate(params, *, unhashable_fallback=True):
     # Weed out strict duplicates, preserving the first of each occurrence.
     try:
-        all_params = set(params)
+        deduped = dict.fromkeys(params)
     except TypeError:
         if not unhashable_fallback:
             raise
@@ -319,15 +319,7 @@ def _deduplicate(params, *, unhashable_fallback=True):
         # Happens for cases like `Annotated[dict, {'x': IntValidator()}]`
         return _deduplicate_unhashable(params)

-    if len(all_params) == len(params):  # fast case
-        return params
-    new_params = []
-    for t in params:
-        if t in all_params:
-            new_params.append(t)
-            all_params.remove(t)
-    assert not all_params, all_params
-    return new_params
+    return deduped
```

File: Lib/typing.py [hunk 9]
```diff
@@ -751,7 +766,9 @@ def open_helper(file: str, mode: MODE) -> str:
     parameters = _flatten_literal_params(parameters)
 
     try:
-        parameters = tuple(p for p, _ in _deduplicate(list(_value_and_type_iter(parameters))))
+        parameters = tuple(
+            p for p, _ in _deduplicate(list(_value_and_type_iter(parameters)))
+        )
```
Comment (AlexWaygood) [inline Lib/typing.py]: There's no change being made to this line now except the formatting; let's keep it as it is to preserve `git blame`:

```suggestion
        parameters = tuple(p for p, _ in _deduplicate(list(_value_and_type_iter(parameters))))
```

File: Lib/typing.py [hunk 10]
```diff
@@ -308,19 +308,34 @@ def _unpack_args(args):
...
-    return params
-
+    try:
+        deduped = dict.fromkeys(params)
+    except TypeError:
+        if not unhashable_fallback:
+            raise
+        # Happens for cases like `Annotated[dict, {'x': IntValidator()}]`
+        return _deduplicate_unhashable(params)
+    return deduped
```
Comment (AlexWaygood) [inline Lib/typing.py]:
```suggestion
        return dict.fromkeys(params)
    except TypeError:
        if not unhashable_fallback:
            raise
        # Happens for cases like `Annotated[dict, {'x': IntValidator()}]`
        return _deduplicate_unhashable(params)
```

File: Lib/test/test_typing.py [hunk 11]
```diff
@@ -8287,6 +8368,11 @@ def test_hash_eq(self):
             {Annotated[int, 4, 5], Annotated[int, 4, 5], Annotated[T, 4, 5]},
             {Annotated[int, 4, 5], Annotated[T, 4, 5]}
         )
+        # Unhashable `metadata` raises `TypeError`:
+        self.assertRaises(TypeError, hash, Annotated[int, []])
+        class A:
+            __hash__ = None
+        self.assertRaises(TypeError, hash, Annotated[int, A()])
```
Comment (AlexWaygood) [inline Lib/test/test_typing.py]: I'd prefer to use the contextmanager form here; I find it more readable :)

```suggestion
        with self.assertRaises(TypeError):
            hash(Annotated[int, []])

        class A:
            __hash__ = None
        with self.assertRaises(TypeError):
            hash(Annotated[int, A()]))
```

File: Lib/test/test_types.py [hunk 12]
```diff
@@ -711,6 +711,18 @@ def test_hash(self):
...
+        class UnhashableMeta(type):
+            __hash__ = None
+
+        class A(metaclass=UnhashableMeta): ...
+        class B(metaclass=UnhashableMeta): ...
+
+        self.assertEqual((A | B).__args__, (A, B))
+        self.assertRaises(TypeError, hash, A | B)
+        self.assertRaises(TypeError, hash, int | B)
+        self.assertRaises(TypeError, hash, A | int)
```
Comment (AlexWaygood) [inline Lib/test/test_types.py]: Make sure it's the `hash()` call that's causing the TypeError to be raised here, not the union operation itself:

```suggestion
        union1 = A | B
        with self.assertRaises(TypeError):
            hash(union1)

        union2 = int | B
        with self.assertRaises(TypeError):
            hash(union2)

        union3 = A | int
        with self.assertRaises(TypeError):
            hash(union3)
```

File: Lib/test/test_typing.py [hunk 13]
```diff
@@ -1769,6 +1770,18 @@ def test_union_union(self):
...
+        class UnhashableMeta(type):
+            __hash__ = None
+
+        class A(metaclass=UnhashableMeta): ...
+        class B(metaclass=UnhashableMeta): ...
+
+        self.assertEqual(Union[A, B].__args__, (A, B))
+        self.assertRaises(TypeError, hash, Union[A, B])
+        self.assertRaises(TypeError, hash, Union[int, B])
+        self.assertRaises(TypeError, hash, Union[A, int])
```
Comment (AlexWaygood) [inline Lib/test/test_typing.py]:
```suggestion
        union1 = Union[A, B]
        with self.assertRaises(TypeError):
            hash(union1)

        union2 = Union[int, B]
        with self.assertRaises(TypeError):
            hash(union2)

        union3 = Union[A, int]
        with self.assertRaises(TypeError):
            hash(union3)
```

File: Lib/test/test_typing.py [hunk 14]
```diff
@@ -8287,6 +8368,14 @@ def test_hash_eq(self):
...
             {Annotated[int, 4, 5], Annotated[T, 4, 5]}
         )
+        # Unhashable `metadata` raises `TypeError`:
+        with self.assertRaises(TypeError):
+            hash(Annotated[int, []])
+
+        class A:
+            __hash__ = None
+        with self.assertRaises(TypeError):
+            hash(Annotated[int, A()]))
```
Comment (AlexWaygood) [inline Lib/test/test_typing.py]:
```suggestion
        a1 = Annotated[int, []]
        with self.assertRaises(TypeError):
            hash(a1)

        class A:
            __hash__ = None
        a2 = Annotated[int, A()]
        with self.assertRaises(TypeError):
            hash(a2)
```

**Comment by sobolevn** [comment]
We just did 6 rounds of nitpicks :)

File: Lib/test/test_typing.py [hunk 15]
```diff
@@ -8267,6 +8286,76 @@ def test_flatten(self):
...
+        class ValueRange:
+            lo: int
+            hi: int
+        v = ValueRange(1, 2)
+        self.assertEqual(get_args(Annotated[int, v] | None),
+                         (Annotated[int, v], type(None)))
+        self.assertEqual(get_args(Union[Annotated[int, v], None]),
+                         (Annotated[int, v], type(None)))
+        self.assertEqual(get_args(Optional[Annotated[int, v]]),
+                         (Annotated[int, v], type(None)))
```
Comment (AlexWaygood) [inline Lib/test/test_typing.py]:
```suggestion
                         (Annotated[int, v], types.NoneType))
        self.assertEqual(get_args(Union[Annotated[int, v], None]),
                         (Annotated[int, v], types.NoneType))
        self.assertEqual(get_args(Optional[Annotated[int, v]]),
                         (Annotated[int, v], types.NoneType))
```

**Review by AlexWaygood — APPROVED** [review]
LGTM. @JelleZijlstra, anything more from you?

**Review by JelleZijlstra — APPROVED** [review]

**Comment by JelleZijlstra** [comment]
Thanks, and sorry for the wait!

Do you think this should be backported? I lean yes because the old behavior feels like a bug.

**Comment by AlexWaygood** [comment]
> Do you think this should be backported? I lean yes because the old behavior feels like a bug.

Yeah, agreed

**Pull request merged**